src/app/clusters/concentration-measurement-server/concentration-measurement-server.h

```diff
@@ -0,0 +1,647 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "concentration-measurement-cluster-objects.h"
+#include <app/AttributeAccessInterface.h>
+#include <app/data-model/Nullable.h>
+#include <app/reporting/reporting.h>
+#include <app/util/af.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/IntrusiveList.h>
+#include <type_traits>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ConcentrationMeasurement {
+
+struct DummyType
+{
+};
+
+/**
+ * This class provides the base implementation for the server side of the Concentration Measurement cluster as well as an API for
+ * setting the values of the attributes.
+ *
+ * @tparam NumericMeasurementEnabled whether the cluster supports numeric measurement
+ * @tparam LevelIndicationEnabled whether the cluster supports level indication
+ * @tparam MediumLevelEnabled whether the Level Indication Feature supports medium level
+ * @tparam CriticalLevelEnabled whether the Level Indication Feature supports critical level
+ * @tparam PeakMeasurementEnabled whether the Numeric Measurement Feature supports peak measurement
+ * @tparam AverageMeasurementEnabled whether the Numeric Measurement Feature supports average measurement
+ */
+template <bool NumericMeasurementEnabled, bool LevelIndicationEnabled, bool MediumLevelEnabled, bool CriticalLevelEnabled,
+          bool PeakMeasurementEnabled, bool AverageMeasurementEnabled>
+class Instance : public AttributeAccessInterface
+{
+private:
+    static const int WINDOW_MAX = 604800;
+
+    EndpointId mEndpointId{};
+    ClusterId mClusterId{};
+
+    // Attribute data store
+    MeasurementMediumEnum mMeasurementMedium;
+    std::conditional_t<NumericMeasurementEnabled, DataModel::Nullable<float>, DummyType> mMeasuredValue;
```

Comment:
Does this actually make mMeasuredValue take 0 space?

Reply:
no, it makes it take space of `DummyType` which is likley around a byte. 

I tried for days to get `enable_if_t` working here but I couldn't get it to work - likely a lack of understanding/knowledge on my part though.

Reply:
For that second bit: I am not following.  When I do:
```
    template <bool Enabled = AverageMeasurementEnabled, typename = std::enable_if_t<Enabled, CHIP_ERROR>>
    CHIP_ERROR SetAverageMeasuredValue(DataModel::Nullable<float> aAverageMeasuredValue)
```

and change all-clusters-app to use `CreateNumericMeasurementAndLevelIndicationConcentrationCluster<true, true, true, false>` for one of the instances, I get:
```
error: no matching member function for call to 'SetAverageMeasuredValue'
    gCarbonDioxideConcentrationMeasurementInstance.SetAverageMeasuredValue(MakeNullable(1.0f));
```
as expected, and if I comment out that one line calling the disabled function, things compile.

For the first bit, what you could do is something like:

```
namespace detail {
struct DummyType1
{
};
struct DummyType2
{
};
// etc

class AverageMembers {
protected:
    DataModel::Nullable<float> mAverageMeasuredValue;
    uint32_t mAverageMeasuredValueWindow;
};
class PeakMembers {
protected:
    DataModel::Nullable<float> mPeakMeasuredValue;
    uint32_t mPeakMeasuredValueWindow;
};
// etc
}
...
class Instance : public AttributeAccessInterface, protected std::conditional_t<AverageMeasurementEnabled, detail::AverageMembers, DummyType1>,
                 protected std::conditional_t<PeakMeasurementEnabled, detail::PeakMembers, DummyType2>
```

and then use `this->mFoo` when referring to the members in the class, because their existence will be template-arg-dependent, so the compiler needs to be told you mean the member.

But it's worth checking whether this is worth it, in that when things _are_ enabled you might not get very good packing and end up with extra padding bits.  But on the other hand, right now the ordering has those padding bits anyway, so it would certainly be no worse than what we have now.